AppCenterIdentity/AppCenterIdentity/MSIdentity.m

```diff
@@ -310,4 +305,62 @@ - (void)removeAuthToken {
   }
 }
 
+- (void)acquireTokenSilentlyWithMSALAccount:(MSALAccount *)account {
+  __weak typeof(self) weakSelf = self;
+  [self.clientApplication
+      acquireTokenSilentForScopes:@[ (NSString * _Nonnull) self.identityConfig.identityScope ]
+                          account:account
+                  completionBlock:^(MSALResult *result, NSError *e) {
+                    typeof(self) strongSelf = weakSelf;
+                    if (e) {
+                      MSLogWarning([MSIdentity logTag],
+                                   @"Silent acquisition of token failed with error: %@. Triggering interactive acquisition", e);
+                      [strongSelf acquireTokenInteractively];
+                    } else {
+                      [MSAuthTokenContext sharedInstance].authToken = result.idToken;
+                      [strongSelf saveAuthToken:result.idToken];
+                      [strongSelf saveAccountId:(NSString * _Nonnull) result.account.homeAccountId.identifier];
+                    }
+                  }];
+}
+
+- (void)acquireTokenInteractively {
+  __weak typeof(self) weakSelf = self;
+  [self.clientApplication acquireTokenForScopes:@[ (NSString * _Nonnull) self.identityConfig.identityScope ]
+                                completionBlock:^(MSALResult *result, NSError *e) {
+                                  if (e) {
+                                    MSLogError([MSIdentity logTag], @"User sign-in failed. Error: %@", e);
+                                  } else {
+                                    typeof(self) strongSelf = weakSelf;
+                                    [MSAuthTokenContext sharedInstance].authToken = result.idToken;
+                                    [strongSelf saveAuthToken:result.idToken];
+                                    [strongSelf saveAccountId:(NSString * _Nonnull) result.account.homeAccountId.identifier];
+                                  }
+                                }];
+}
+
+- (MSALAccount *)retrieveAccountWithAccountId:(NSString *)homeAccountId {
+  if (!homeAccountId) {
+    return nil;
+  }
+  NSError *error;
+  MSALAccount *account = [self.clientApplication accountForHomeAccountId:homeAccountId error:&error];
+  if (error) {
+    MSLogWarning([MSIdentity logTag], @"Could not get MSALAccount for homeAccountId.");
```

Comment:
Should we print the error message from the error?

Reply:
The scenario is if we get an error from `accountForHomeAccountId:homeAccountId:` from MSAL API, which means we won't be able to use the persisted `accountId` to acquire token silently from MSAL, we will "fall back" to interactive login.

It is not an error/failure case, therefore it shouldn't be an error log. Warning level should be more appropriate.

Reply:
True I agree with that but I was more talking about should we print the message from the error object that we get from MSAL inside our warning log :sweat_smile:. That wasn't clear sorry fro the confusion.

Reply:
Sorry I misread that. It's fixed now.